rclpy/rclpy/duration.py

```diff
@@ -16,10 +16,20 @@
 from rclpy.impl.implementation_singleton import rclpy_implementation as _rclpy
 
 
+CONVERSION_CONSTANT = 10 ** 9
+
+
 class Duration:
+    """A period between two time points, with nanosecond precision."""
 
     def __init__(self, *, seconds=0, nanoseconds=0):
-        total_nanoseconds = int(seconds * 1e9)
+        """
+        Create an instance of :class:`Duration`, combined from given seconds and nanoseconds.
+
+        :param seconds: Time span seconds, if any, fractional part will be discarded.
+        :param nanoseconds: Time span nanoseconds, if any, fractional part will be discarded.
+        """
+        total_nanoseconds = int(seconds) * CONVERSION_CONSTANT
```

Comment:
This causes test failures like https://build.ros2.org/job/Rpr__rclpy__ubuntu_focal_amd64/674/testReport/junit/rclpy.rclpy.test.test_time/TestTime/test_duration_construction/

How about:
```suggestion
        total_nanoseconds = int(seconds * CONVERSION_CONSTANT)
```

This discards the fractional part after converting to nanoseconds, rather than before. I'm not sure if this is the intended semantics though.

Reply:
Made the logic mirror that in `rclpy.Time`. 

The semantics of it are a bit questionable, though. If `rclcpp` can be taken as the reference implementation, then there the [corresponding constructor](https://github.com/ros2/rclcpp/blob/master/rclcpp/src/rclcpp/duration.cpp#L34) accepts only integer values. And in general it's a bad idea to use floats for sub-second precision. 

But that would go beyond this bugfix. 

Some tests, however, still fail. Looks like parts of the stack might already be coded to expect this specific floating-point behavior..

Reply:
Yeah, I don't think the change I suggested is the way to go, but also as you mentioned other parts of the stack expect specific floating-point behavior.

Reply:
The problem with the failing test is that it's poorly written, very much on the edge of failing due to sub-millisecond timing differences between different test runs. Some test runs like https://build.ros2.org/job/Rpr__rclpy__ubuntu_focal_amd64/677/ succeeded despite no functional differences in code. And looking at builds for other pull requests, this test fails frequently, for no reasons connected with changes made. 

I made another change - discovered and made use of https://github.com/ros2/rclpy/blob/master/rclpy/rclpy/constants.py. No functional difference, but now all tests succeed. 

Do you think I should rewrite the test to be less sensitive to random timing differences? I would assert the length of callbacks in https://github.com/ros2/rclpy/blob/master/rclpy/test/test_timer.py#L104 to be in `(2, 3)`.

Reply:
> The problem with the failing test is that it's poorly written, very much on the edge of failing due to sub-millisecond timing differences between different test runs

I agree that the nature of the test failures does not always reflect proposed changes.


> Do you think I should rewrite the test to be less sensitive to random timing differences? I would assert the length of callbacks in https://github.com/ros2/rclpy/blob/master/rclpy/test/test_timer.py#L104 to be in `(2, 3)`.

That's not a bad idea, but I reckon you should file a separate issue about the tests (if for nothing else but visibility) and then make PRs to close said issue. That'll also give a chance for maintainers to share their opinion about the timing tests.

Reply:
>  > Do you think I should rewrite the test to be less sensitive to random timing differences? I would assert the length of callbacks in https://github.com/ros2/rclpy/blob/master/rclpy/test/test_timer.py#L104 to be in (2, 3).

> That's not a bad idea, but I reckon you should file a separate issue about the tests (if for nothing else but visibility) and then make PRs to close said issue. That'll also give a chance for maintainers to share their opinion about the timing tests.

Agreed.  We would very much appreciate a new issue and a new PR to make the timing tests more reliable.  Thanks!